sematic/plugins/external_resource/ray/cluster.py

```diff
@@ -0,0 +1,387 @@
+# Standard Library
+import logging
+import time
+from dataclasses import dataclass, field, replace
+from typing import Optional, Tuple, Type
+
+# Third-party
+import kubernetes
+from kubernetes.client.rest import ApiException  # type: ignore
+
+# Sematic
+from sematic.abstract_plugin import PluginScope
+from sematic.config.server_settings import ServerSettingsVar, get_server_setting
+from sematic.config.settings import get_active_plugins
+from sematic.db.queries import get_run, get_run_ids_for_resource
+from sematic.plugins.abstract_external_resource import (
+    AbstractExternalResource,
+    ManagedBy,
+    ResourceState,
+)
+from sematic.plugins.abstract_kuberay_wrapper import (
+    AbstractKuberayWrapper,
+    RayClusterConfig,
+)
+from sematic.plugins.kuberay_wrapper.standard import StandardKuberayWrapper
+from sematic.scheduling.kubernetes import load_kube_config
+from sematic.utils.exceptions import UnsupportedUsageError
+
+try:
+    # Third-party
+    import ray  # type: ignore
+    from ray.exceptions import GetTimeoutError  # type: ignore
+except ImportError as e:
+    raise ImportError(
+        "RayCluster can only be used in Sematic if your code has a dependency on Ray"
+    ) from e
+
+
+logger = logging.getLogger(__name__)
+
+
+_VALIDATION_INT = 1
+
+
+def _no_default_cluster() -> RayClusterConfig:
+    raise ValueError(
+        f"RayCluster must be initialized with a {RayClusterConfig.__name__} "
+        "for 'cluster'"
+    )
+
+
+@dataclass(frozen=True)
+class RayCluster(AbstractExternalResource):
+
+    # Since the parent class has defaults, all params here must technically
+    # have defaults. Here we raise an error if no value is provided though.
+    config: RayClusterConfig = field(default_factory=_no_default_cluster)
+    _cluster_name: Optional[str] = None
+    _head_uri: Optional[str] = None
+    _deleted_cluster: bool = False
+
+    @classmethod
+    def _kuberay_wrapper(cls) -> Type[AbstractKuberayWrapper]:
+        plugins = get_active_plugins(
+            PluginScope.KUBERAY, default=[StandardKuberayWrapper]
+        )
```

Comment:
You may need to `typing.cast` here as `get_active_plugins` likely returns `List[AbstractPlugin]`.